bundle/artifacts/artifacts.go

```diff
@@ -0,0 +1,143 @@
+}
+
+var uploadMutators map[config.ArtifactType]mutatorFactory = map[config.ArtifactType]mutatorFactory{}
+
+func getBuildMutator(t config.ArtifactType, name string) bundle.Mutator {
+	mutatorFactory, ok := buildMutators[t]
+	if !ok {
+		mutatorFactory = BasicBuild
+	}
+
+	return mutatorFactory(name)
+}
+
+func getUploadMutator(t config.ArtifactType, name string) bundle.Mutator {
+	mutatorFactory, ok := uploadMutators[t]
+	if !ok {
+		mutatorFactory = BasicUpload
+	}
+
+	return mutatorFactory(name)
+}
+
+// Basic Build defines a general build mutator which builds artifact based on artifact.BuildCommand
+type basicBuild struct {
+	name string
+}
+
+func BasicBuild(name string) bundle.Mutator {
+	return &basicBuild{name: name}
+}
+
+func (m *basicBuild) Name() string {
+	return fmt.Sprintf("artifacts.Build(%s)", m.name)
+}
+
+func (m *basicBuild) Apply(ctx context.Context, b *bundle.Bundle) error {
+	artifact, ok := b.Config.Artifacts[m.name]
+	if !ok {
+		return fmt.Errorf("artifact doesn't exist: %s", m.name)
+	}
+
+	cmdio.LogString(ctx, fmt.Sprintf("artifacts.Build(%s): Building...", m.name))
+
+	out, err := artifact.Build(ctx)
+	if err != nil {
+		return fmt.Errorf("artifacts.Build(%s): %w, output: %s", m.name, err, out)
+	}
+	cmdio.LogString(ctx, fmt.Sprintf("artifacts.Build(%s): Build succeeded", m.name))
+
+	return nil
+}
+
+// Basic Upload defines a general upload mutator which uploads artifact as a library to workspace
+type basicUpload struct {
+	name string
+}
+
+func BasicUpload(name string) bundle.Mutator {
+	return &basicUpload{name: name}
+}
+
+func (m *basicUpload) Name() string {
+	return fmt.Sprintf("artifacts.Build(%s)", m.name)
+}
+
+func (m *basicUpload) Apply(ctx context.Context, b *bundle.Bundle) error {
+	artifact, ok := b.Config.Artifacts[m.name]
+	if !ok {
+		return fmt.Errorf("artifact doesn't exist: %s", m.name)
+	}
+
+	if artifact.File == "" {
+		return fmt.Errorf("artifact source is not configured: %s", m.name)
+	}
+
+	cmdio.LogString(ctx, fmt.Sprintf("artifacts.Upload(%s): Uploading...", m.name))
+
+	r, err := uploadArtifact(ctx, artifact, b)
+	if err != nil {
+		return fmt.Errorf("artifacts.Upload(%s): %w", m.name, err)
+	}
+
+	artifact.RemotePath = r
+	cmdio.LogString(ctx, fmt.Sprintf("artifacts.Upload(%s): Upload succeeded", m.name))
+	return nil
+}
+
+// Function to upload artifact as a library to Workspace
+// Currenly it does not work correctly because Workspace.Import API can not import libraries
+// but only notebooks or files
+func uploadArtifact(ctx context.Context, a *config.Artifact, b *bundle.Bundle) (string, error) {
+	raw, err := os.ReadFile(a.File)
+	if err != nil {
+		return "", fmt.Errorf("unable to read %s: %w", a.File, errors.Unwrap(err))
+	}
+
+	artifactPath := b.Config.Workspace.ArtifactsPath
+	if artifactPath == "" {
+		return "", fmt.Errorf("remote artifact path not configured")
+	}
+
+	remotePath := path.Join("/Workspace", artifactPath, path.Base(a.File))
```

Comment:
This should not be necessary. The paths under `b.Config.Workspace` should be usable as is.

Only when you read this file through FUSE do you need a POSIX-valid path, which for the workspace filesystem is mounted under `/Workspace`.